wgsl/index.bs

```diff
@@ -9597,29 +9899,50 @@ value|non-uniform=] during the execution of the function call.
         create a <dfn noexport>Value_return_i</dfn> node.
+        This represents the function's effect on the uniformity of the value pointed to by the *i*<sup>th</sup> parameter.
 * Walk over the syntax of the function, adding nodes and edges to the graph following the rules of the next sections ([[#func-var-value-analysis]], [[#uniformity-statements]], [[#uniformity-function-calls]], [[#uniformity-expressions]]), using [=CF_start=] as the starting control-flow for the function's body.
-* For each [=Value_return_i=] node, record which [=param_i=] nodes are reachable from it.
-* Look at which nodes are reachable from [=RequiredToBeUniform=].
-    * If this set includes the node [=MayBeNonUniform=], then reject the program.
-    * If this set includes [=CF_start=], then the [=call site tag=] for the function is [=CallSiteRequiredToBeUniform=].
-    * Otherwise, the [=call site tag=] is [=CallSiteNoRestriction=].
-    * For each [=param_i=] in this set, the corresponding [=parameter tag=] is [=ParameterRequiredToBeUniform=].
-    * Remove from the graph all nodes that have been visited.
-* If [=Value_return=] exists, look at which nodes are reachable from it
-    * If this set includes [=MayBeNonUniform=], then the [=function tag=] is [=ReturnValueMayBeNonUniform=].
-    * For each [=param_i=] in this set, the corresponding [=parameter tag=] is [=ParameterRequiredToBeUniformForReturnValue=].
-* For each [=Value_return_i=] node, look at which nodes are reachable from it
-    * If this set includes [=MayBeNonUniform=], the corresponding [=pointer parameter tag=] is [=PointerParameterMayBeNonUniform=].
-    * Otherwise, the corresponding [=pointer parameter tag=] is [=PointerParameterNoRestriction=].
-* If the [=function tag=] has not been assigned, then it is [=NoRestriction=].
-* For each parameter, if it has not been assigned a [=parameter tag=], then it is [=ParameterNoRestriction=].
-
-Note: The entire graph can be destroyed at this point. The tags listed above are all that we need to remember to analyze callers of this function.
+* Initialize as follows:
+    * The [=function tag=] is initialized to [=NoRestriction=].
+    * The [=call site tag=] is initialized to [=CallSiteNoRestriction=].
+    * The [=parameter tag=] for each [=param_i=] is initialized to [=ParameterNoRestriction=].
+    * The [=pointer parameter tag=] for each [=param_i=], if it exists, is initialized to [=PointerParameterNoRestriction=].
+* For each severity *S* in the order {[=severity/error=], [=severity/warning=], [=severity/info=]}, perform the following:
+    * Let *R.S* be the set of nodes reachable from [=RequiredToBeUniform.S=].
+    * Let *W* be the [=witness set=] associated with [=RequiredToBeUniform.S=].
+    * If *R.S* includes the node [=MayBeNonUniform=], then trigger a [=uniformity failure=]:
+         * For each [=witness=] *w* in *W*, [=triggered|trigger=] a diagnostic with severity *S*, using the call site and triggering rule from *w*.
```

Comment:
I'm trying to understand why we need witness sets at all, now that we explicitly encode the diagnostic severity in the graph. I think this line is the only place where they have an effect, and that effect is to propagate the triggering rule to the diagnostic that is raised (i.e. `derivative_uniformity` vs unnamed).

The triggering rule is used to determine the diagnostic severity, but we've already figured that out at the callsite and propagated it through the graph. So the witness set machinery feels redundant to me, and it adds additional complexity to the analysis that does not reflect the implementation.

Please let me know if I'm missing something here.

Reply:
>  I think this line is the only place where they have an effect,

Yes.

> and that effect is to propagate the triggering rule to the diagnostic that is raised (i.e. derivative_uniformity vs unnamed).

Not quite.  We have to create diagnostics, and a triggering location is one of the 3 defining properties of a diagnostic.  So we needed the source location, and that has to be propagated up.  "Witness" = (source location, severity).

Funny thing.  I had deleted the witness machinery, and for a while I had "TODO"s to track locations, then added them back exactly for this reason.

Reply:
The witness set exists to propagate the triggering location out of a function. It still seems required to me.